File: Assignment/Ass2/Q2/A2_P2_123020227.py
class Node:
    def __init__(self):
        self.num = 0
        self.data = []
    def add(self, element):
        self.data.append(element)

def multiply_rows(row_a_node, matrix_b):
    result = {}
    for col_a, val_a in row_a_node.data:
        row_b_node = matrix_b[col_a]
        for col_b, val_b in row_b_node.data:
            if col_b not in result:
                result[col_b] = 0
            result[col_b] += val_a * val_b
    result_list = []
    for col in sorted(result.keys()): 
        if result[col] != 0:
            result_list.append([col, result[col]])
    return result_list

def main():
    n, m = map(int, input().split())
    matrix_a = [None] * n
    matrix_b = [None] * m
    for i in range(n):
        line = input().split()
        node = Node()
        num = int(line[0])
        node.num = num
        for j in range(num):
            col = int(line[1 + 2 * j])
            val = int(line[2 + 2 * j])
            node.add([col, val])
        matrix_a[i] = node
    input()
    m, p = map(int, input().split())
    for i in range(m):
        line = input().split()
        node = Node()
        num = int(line[0])
        node.num = num
        for j in range(num):
            col = int(line[1 + 2 * j])
            val = int(line[2 + 2 * j])
            node.add([col, val])
        matrix_b[i] = node

    matrixC = ""
    front = 0
    for i in range(n):
        row_a_node = matrix_a[i]
        result_data = multiply_rows(row_a_node, matrix_b, m, p)
        row_string = str(len(result_data))
        for col, val in result_data:
            row_string += f" {col} {val}"
        matrixC += row_string + "\n"
    print(f"{n} {p}")
    print(matrixC, end="")

if __name__ == "__main__":
    main()

File: Assignment/Ass2/A2_P2_123020227.java
import java.util.*;

class Node {
    int num;
    List<int[]> data;

    public Node() {
        this.num = 0;
        this.data = new ArrayList<>();
    }

    public void add(int[] element) {
        this.data.add(element);
    }
}

public class Main {

    public static List<int[]> multiplyRows(Node rowA, List<Node> matrixB) {
        Map<Integer, Integer> result = new TreeMap<>();
        
        for (int[] entryA : rowA.data) {
            int colA = entryA[0];
            int valA = entryA[1];
            
            Node rowB = matrixB.get(colA);
            
            for (int[] entryB : rowB.data) {
                int colB = entryB[0];
                int valB = entryB[1];
                
                result.put(colB, result.getOrDefault(colB, 0) + valA * valB);
            }
        }

        List<int[]> resultList = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : result.entrySet()) {
            if (entry.getValue() != 0) {
                resultList.add(new int[] {entry.getKey(), entry.getValue()});
            }
        }
        return resultList;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        int m = scanner.nextInt();

        List<Node> matrixA = new ArrayList<>(n);

        for (int i = 0; i < n; i++) {
            Node node = new Node();
            int num = scanner.nextInt();
            node.num = num;
            for (int j = 0; j < num; j++) {
                int col = scanner.nextInt();
                int val = scanner.nextInt();
                node.add(new int[] {col, val});
            }
            matrixA.add(node);
        }

        scanner.nextLine();

        m = scanner.nextInt();
        int p = scanner.nextInt();
        List<Node> matrixB = new ArrayList<>(m);
        for (int i = 0; i < m; i++) {
            Node node = new Node();
            int num = scanner.nextInt();
            node.num = num;
            for (int j = 0; j < num; j++) {
                int col = scanner.nextInt();
                int val = scanner.nextInt();
                node.add(new int[] {col, val});
            }
            matrixB.add(node);
        }
        StringBuilder matrixC = new StringBuilder();
        for (int i = 0; i < n; i++) {
            Node rowA = matrixA.get(i);
            List<int[]> resultData = multiplyRows(rowA, matrixB);
            
            StringBuilder rowString = new StringBuilder();
            rowString.append(resultData.size());
            for (int[] entry : resultData) {
                rowString.append(" ").append(entry[0]).append(" ").append(entry[1]);
            }
            matrixC.append(rowString).append("\n");
        }
        System.out.println(n + " " + p);
        System.out.print(matrixC.toString());

        scanner.close();
    }
}
